client-src/elements/chromedash-app.js

```diff
@@ -138,99 +138,177 @@ class ChromedashApp extends LitElement {
-      // but instead just scroll to the element identified by the hash.
+    // Whether any changes have been made to form fields.
+    // Undo of changes does not undo this setting.
+    // This is null only when "loading" a new page.
+    let changesMade = null;
+
+    let beforeUnloadHandler = null;
+
+    const resetBeforeUnloadHandler = () => {
+      if (!beforeUnloadHandler) return;
+      // Remove previous beforeunload handler, to forget changes.
+      window.removeEventListener('beforeunload', beforeUnloadHandler);
+      beforeUnloadHandler = null;
+    };
+
+    // Maybe set up new page, or if the URL is the same, we stay.
+    // Returns true if we are proceeding to the new page, false otherwise.
+    const setupNewPage = (ctx, componentName) => {
+      // If current page is ctx.path and a ctx.hash exists,
+      // don't create a new element but instead
+      // just scroll to the element identified by the hash.
+      // Note, this ignores any query string.
       if (this.currentPage == ctx.path && ctx.hash) {
         if (window.scrollToElement) {
           window.scrollToElement(`#${ctx.hash}`);
         }
-        return true;
+        return false;
+      }
+
+      // If there was a previous page, check if we would lose unsaved changes.
+      if (this.pageComponent) {
+        // Act like we are unloading previous page and loading a new page.
+        if (changesMade) {
+          // Should we use shoelace dialog instead?
+          if (!confirm('You will lose unsaved changes.  Proceed anyway?')) {
+            // Set ctx.handled to false, so we don't change browser's history.
+            ctx.handled = false;
+            return false;
+          }
+        }
       }
+
+      // Loading new page.
+      this.pageComponent = document.createElement(componentName);
+      changesMade = null;
+      resetBeforeUnloadHandler();
+
+      // If anything has changed, set up beforeunload handler.
+      this.pageComponent.addEventListener('sl-change', () => {
+        changesMade = true;
+        // Make sure the beforeunload event handler has been set up.
+        if (beforeUnloadHandler) return;
+        beforeUnloadHandler = (event) => {
+          // Cancel the event, which asks user whether to stay.
+          event.preventDefault();
+          // Chrome requires returnValue to be set.
+          event.returnValue = `You made changes that have not been saved.
+          Are you sure you want to leave?`;
+        };
+        window.addEventListener('beforeunload', beforeUnloadHandler);
+      });
+
+      window.setTimeout(() => {
+        // Allow submit button to proceed, if the form is valid.
+        // The submit button is not necessarily loaded yet, hence the timeout.
+        // Note: in general, there are several more ways to submit a form.
+        const submitButton =
+          this.pageComponent.shadowRoot.querySelector('input[type="submit"');
```

Comment:
What about submit buttons that are in the sidebar?
Do we have any submit buttons that are nested inside multiple levels of web components?
What about the Cancel button?

Reply:
Are the sidebar submit buttons for the same form, or a different form?  I'm looking in the pageComponent only, so far.

There are, in general, other ways to submit the form, which are not handled here.  But maybe handling the 'submit' event for the form will deal with all of them.  Not sure.  

I though about the Cancel button, but actually, it seems like a good idea to warn the user that they are losing changes, in case they did not intend to click the cancel button.  Otherwise, it is way too easy to lose everything by clicking on the cancel button which is right next to the submit button.  There's a meme about that :)

Reply:
BTW, I changed from listening to 'click' on the Submit button to listening to 'submit' on the form.  This catches all ways the form might be submitted.